### PR TITLE
bug(GRW-890): apollo cache override

### PR DESCRIPTION
**Pull request by guilhermespopolin**
## What?

* Provide to _apollo_ InMemoryCache a way to generate an id for `QuoteBundle` types: it relies on quote ids;
* Update _graphql_ quoteCart related queries/mutations so we fetch quote ids while working with that type;

## Why?

`QuoteCart` type [defines a field called `bundle`](https://github.com/HedvigInsurance/giraffe/blob/master/src%2Fresolvers%2Fquote-cart%2FquoteCart.graphqls#L89) which type, [`QuoteBundle`](https://github.com/HedvigInsurance/giraffe/blob/master/src%2Fresolvers%2Fquote-bundle%2FquoteBundle.graphqls#L9), doesn't have a explicit form of identification. As a consequence, we might face some issues with _apollo_ caching as some queries might override the cache which may cause problems. 

To solve this issue we can provide to _apollo_ `InMemoryCache` a way to define a custom id for `QuoteBundle` type and make sure we fetch the necessary data for that id be built while working with that type.

**Ticket(s): [GRW-890](https://hedvig.atlassian.net/browse/GRW-890)**

## Drawbacks

* We need to keep in mind that it's **really important** to request `quotes { id }` while working with `QuoteBundle` type. Otherwise the custom id generation will not work;
* Apollo cache increases significantly  during editions because we don't clear cache stale data while creating a new quote bundle. I've tested with `editQuote` api real quick and it seems this problem doesn't happen with it (which make sense, since we're not creating new quotes but just updating them).

## How test it

You can access the [review app](https://web-onboardi-bug-grw-89-ynytlz.herokuapp.com/se-en/new-member/offer/3934f0ce-5118-4123-8d59-9a5bd28b9673) and check _Cache_ tab into _Apollo Dev Tools_. It should be able to see that `QuoteBundle` types are being normalised:

Screenshot 2022-03-09 at 13 23 59

### Related with

[chore: update due graphl api type changes](https://github.com/HedvigInsurance/web-onboarding/pull/856)

### CODE REVIEW

File: src/client/apolloClient.ts [hunk 1]
```diff
@@ -62,6 +63,20 @@ export const apolloClient = (() => {
...
+          keyFields: (quoteBundle) => {
+            const quotes = (quoteBundle as QuoteBundle).quotes ?? []
+
+            if (quotes.length === 0) return
+
+            // QuoteBundle:<id>+<id>
+            return quotes.reduce(
+              (id, quote, index) =>
+                (id += `${quote.id}${index < quotes.length - 1 ? '+' : ''}`),
+              'QuoteBundle:',
```
Comment (robinandeer) [inline src/client/apolloClient.ts:76]: hehe isn't this overly complicated?

```suggestion
            // QuoteBundle:<id>+<id>
            return `QuoteBundle:${quotes.map((quote) => quote.id).join('+')}`
```

**Review by robinandeer — COMMENTED** [review]
Neat enough solution I think since we are dealing with a resource that's missing an ID which from what I understand is different from how GraphQL is intended to work.

File: src/client/apolloClient.ts (hunk 1 above)
Comment (guilhermespopolin) [reply src/client/apolloClient.ts:76]: OMG hahhaha
So much better

**Comment by guilhermespopolin** [comment]
We decided no go with this solution.

**Pull request closed without merging**